WIP:  Replacing pest with nom

CNLHC opened the pull request.

curquiza commented:
hello @CNLHC
Thanks for your PR! 😁 @irevoire and/or @Kerollmops will review as soon as possible!

irevoire commented:
Actually, I think this PR is not finished yet @curquiza.

@CNLHC if you want to finish this PR I think the next step would be:
1. Fix some edge cases; try to run the tests! It looks like currently, it's not working. When I try to execute a simple filter like `id < 200`. I get this panic: 
```
thread 'tokio-runtime-worker' panicked at 'internal error: entered unreachable code', milli/src/search/facet/filter_condition.rs:177:18
```
2. Remove pest as a dependency and everything related to pest in general
3. Move the parser in an independent file somewhere else, `milli/src/search/facet/filter_parser.rs` would be a good place I guess.
4. Rebase on `main`. This should not be too hard but the parser has been slightly modified with #322.

If you need help with something do not hesitate to contact us :smile: 

And above all, thanks for your work!

curquiza commented:
I converted it to a draft PR since it's not a finished work
@CNLHC can you pass it as ready for review once you've done? 🙂

![Uploading Capture d’écran 2021-09-22 à 13.59.10.png…]()

CNLHC commented:
> I converted it to a draft PR since it's not a finished work
> @CNLHC can you pass it as ready for review once you've done? 
> 
> <img alt="Capture d’écran 2021-09-22 à 13 59 10" width="1066" src="https://user-images.githubusercontent.com/20380692/134339763-7e420388-78c3-4def-97a0-4efe23ba00eb.png">

Ok, i will click this button once i fixed the issues mentioned by @irevoire .

CNLHC commented:
> Fix some edge cases; try to run the tests! It looks like currently, it's not working. When I try to execute a simple filter like id < 200. I get this panic:

It is caused by a typo and is fixed now. I also add a corresponding test case to it.

> Remove pest as a dependency and everything related to pest in general.

removed.

> Move the parser in an independent file somewhere else, milli/src/search/facet/filter_parser.rs would be a good place I guess.

All the parser logic were moved to [milli/src/search/facet/filter_parser.rs](milli/src/search/facet/filter_parser.rs)

> Rebase on main. This should not be too hard but the parser has been slightly modified with Geosearch #322.

I have rebased this branch to the latest `main` and implement parser logic for `_geoRadius`.

irevoire commented:
I think we are almost ready to merge. Can you apply this comment: https://github.com/meilisearch/milli/pull/358#discussion_r727325123 and run rustfmt one last time?

I will let @Kerollmops merge or add comments though.
